Data/Matrix.hs

```diff
@@ -556,6 +556,31 @@ inverse m
             rref'd = rref adjoinedWId
         in rref'd >>= return . submatrix 1 (nrows m) (ncols m + 1) (ncols m * 2)
 
+-- | /O(strassenMixed)/. The inverse of a square matrix.
+--   Inverts blockwise. Significantly faster than naive Gaussian elimination.
+--   Same time complexity as whatever matrix multiplication (*) currently uses.
+--   See https://en.wikipedia.org/wiki/Invertible_matrix#Blockwise_inversion
+blockInverse :: (Fractional a, Eq a) => Matrix a -> Either String (Matrix a)
+blockInverse m
+    | nrows m /= ncols m = inverse m
+    | otherwise = Right $ blockInverseUnsafe m
+
+-- blockInverse that assumes that it is called
+-- on a square matrix.
+blockInverseUnsafe :: (Fractional a, Eq a) => Matrix a -> Matrix a
+blockInverseUnsafe m
+    | nrows m < 10 = either (\_ -> m) id $ inverse m
```

Comment:
I am concerned about speed. Otherwise I wouldn't have implemented several methods for multiplication or determinant calculation. Nevertheless, safety goes first, so issues like #38 are way more important. I think the `inverse` function could use a rewrite, and be at last a function one can trust on. It should return `Left` when the matrix is not invertible and `Right` when it is. It _must_ not throw any exceptions. And it should be as efficient as we can get it to be. Unfortunately, I do not have much time currently to take care of the development of this library. But I still maintain it and review patches. It is sad that there are quite a few really bad bugs around (check the issue list), but time is a limited resource. Although, of course, I am super grateful if you (or anybody else) can take a look at these, and I would be happy to review the code. By the way, thanks for looking at this and trying to improve the inverse function!